native/cocos/audio/oalsoft/AudioEngine-soft.cpp

```diff
@@ -532,3 +540,143 @@ void AudioEngineImpl::uncacheAll() {
+            ret = 1.0f / (float)MAXINT32;
+            break;
+        case cc::AudioDataType::FLOAT_32:
+            //TODO: check the reduceFactor of float 32 to reset [-1,1]
+            ret = 1.0f;
+        default:
+            break;
+    }
+    return ret;
+}
+typedef std::function<float(char *)> AudioDataInterpreter;
+AudioDataInterpreter getFormatedData(AudioDataType type) {
+    AudioDataInterpreter f = nullptr;
+    switch (type) {
+        case cc::AudioDataType::SIGNED_16:
+            f = [](char *buf) -> float { return (float)(*reinterpret_cast<short *>(buf)); };
+            break;
+        case cc::AudioDataType::SIGNED_32:
+            f = [](char *buf) -> float { return (float)(*reinterpret_cast<int32_t *>(buf)); };
+            break;
+        case cc::AudioDataType::FLOAT_32:
+            f = [](char *buf) -> float { return (float)(*reinterpret_cast<float *>(buf)); };
+            break;
+        default:
+            break;
+    }
+    return f;
+}
+
+void AudioEngineImpl::getPCMBuffer(const char *url, uint32_t channelID, ccstd::vector<float> &pcmData) {
+    ccstd::string _fileFullPath = FileUtils::getInstance()->fullPathForFilename(url);
+    if (_fileFullPath == "") {
+        CC_LOG_DEBUG("file %s does not exist or failed to load", url);
+        return;
+    }
+    AudioDecoder *decoder = AudioDecoderManager::createDecoder(_fileFullPath.c_str());
+    if (decoder == nullptr) {
+        CC_LOG_DEBUG("decode %s failed, the file formate might not support", url);
+        return;
+    }
+    do {
+        if (!decoder->open(_fileFullPath.c_str())) {
+            CC_LOG_ERROR("[Audio Decoder] File open failed %s", url);
+            break;
+        }
+        const uint32_t bytesPerFrame = decoder->getBytesPerFrame();
+        const uint32_t bytesPerChannel = decoder->getBytesPerChannel();
+        const uint32_t channelCount = decoder->getChannelCount();
+        if (channelID >= channelCount) {
+            CC_LOG_ERROR("channelID invalid, total channel count is %d but %d is required", channelCount, channelID);
+            break;
+        }
+        uint32_t totalFrames = decoder->getTotalFrames();
+        uint32_t remainingFrames = totalFrames;
+        uint32_t framesRead = 0;
+        uint32_t framesToReadOnce = std::min(totalFrames, static_cast<uint32_t>(decoder->getSampleRate() * QUEUEBUFFER_TIME_STEP * QUEUEBUFFER_NUM));
+        AudioDataType type = decoder->getDataType();
+        float redFac = reduceFactor(type);
+        AudioDataInterpreter interpreter = getFormatedData(type);
+        char *tmpBuf;
+        // For some part of type, we dont't actually know its type. 2 bytes => short, 4 bytes => float/int32.
+        char *tmpBufPerChannel;
+
+        tmpBufPerChannel = static_cast<char *>(malloc(bytesPerChannel + 1)); //NOLINT
+        tmpBufPerChannel[bytesPerChannel] = '\0';
+        pcmData.reserve(totalFrames);
+        tmpBuf = static_cast<char *>(malloc(framesToReadOnce * bytesPerFrame));
```

Comment:
Why not define and assign in the same line, like
```c++
char *tmpBuf = static_cast<char *>(malloc(framesToReadOnce * bytesPerFrame));
```

Reply:
Should clear memory value to 0? Or it will contain invalid value.

Reply:
free at the end。